odoo/addons/base/models/ir_cron.py

```diff
@@ -280,73 +291,108 @@ def _acquire_one_job(cls, cr, job_ids):
-            interval = _intervalTypes[job['interval_type']](job['interval_number'])
-            env = api.Environment(job_cr, job['user_id'], {'lastcall': lastcall})
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
             ir_cron = env[cls._name]
+            completed = True
+            for i in range(MAX_BATCH_PER_CRON_JOB):
+                progress = ir_cron._create_progress(job['id'], timed_out=True)
+                job_cr.commit()
+
+                ir_cron_w_context = ir_cron.with_context(ir_cron_progress_id=progress.id)
+                try:
+                    (done, remaining) = ir_cron_w_context._callback(
+                        job['cron_name'], job['ir_actions_server_id'], job['id']
+                    )
+                except Exception:
+                    failed = not progress.done  # If some progress has been committed, we do not consider it a failure
+                else:
+                    failed = False
+                    ir_cron_w_context._notify_progress(done, remaining)
+                finally:
+                    progress.timed_out = False
+                    job_cr.commit()
+
+                completed = not progress.remaining
+                if completed:
+                    break
+                _logger.info('Job `%s`: %s records processed and %s records remaining.',
+                             job['cron_name'], progress.done, progress.remaining)
+
+        return failed, completed
 
-            # Use the user's timezone to compare and compute datetimes,
-            # otherwise unexpected results may appear. For instance, adding
-            # 1 month in UTC to July 1st at midnight in GMT+2 gives July 30
-            # instead of August 1st!
-            now = fields.Datetime.context_timestamp(ir_cron, datetime.utcnow())
-            past_nextcall = fields.Datetime.context_timestamp(
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
+    @classmethod
+    def _process_job(cls, db, cron_cr, job):
+        """ Execute a cron job and re-schedule a call for later if
+        there are remaining records to process. """
```

Comment:
Another docstring suggestion :)

> Execute the cron's server action in a dedicated transaction.
>
> The server action can report progression using the `ir.cron.progress` model. In case it reports that not all records could have been processed this time, then the cron re-triggers itself to be executed again as soon as possible and process the remaining records. Otherwise, when the section action reported that all records have been processed, the cron is scheduled again in the future using its interval configuration.
>
> In case all executions of the cron fail in a period of {MIN_DELTA_BEFORE_DEACTIVATION} and that there are at least {MIN_FAILURES_BEFORE_DEACTIVATION} failures then the cron is automatically deactivated and the admin notified.
>
> The cron is only run once no matter how many intervals and triggers were missed.